README.md
# This is FramerD

**FramerD** is a production-ready platform and database for
applications and services which leverage knowledge and context to
deliver personalized results and responses.

*FramerD* consists of two main components: 

* a **portable distributed object-oriented database** designed to
  support the maintenance and sharing of knowledge bases. Unlike other
  object-oriented databases, FramerD is optimized for the sort of
  pointer-intensive data structures used by semantic networks, frame
  systems, and many intelligent agent applications.

* a flexible and performant scripting language based on **Scheme**
  with special provisions for *Prolog*-style non-deterministic
  programming, advanced text processing, and web application
  development.

FramerD databases readily include millions of searchable frames and
may be distributed over multiple networked machines.

FramerD is implemented in ANSI C for Unix-based platforms including
Linux and macOS.  In addition, Java libraries exist for accessing
FramerD databases and services.

FramerD was originally developed at
[MIT's Media Laboratory](www.media.mit.edu) by Ken Haase. The
laboratory's support, especially from its *News In the Future*
program, is gratefully acknowledged. The primary development of
FramerD is currently supported by
[beingmeta](https://www.beingmeta.com/), a developer of
next-generation content and knowledge management tools.



debian/changelog
framerd (3.9.1-13) unstable; urgency=medium

  * Fix packaging descriptions
  * git patch version framerd-3.9.1-31-g2af7e18

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 11:11:23 -0400

framerd (3.9.1-12) unstable; urgency=medium

  * Split out framerd into framerd-core and framerd-builtin (which provides framerd-modules)
  * git patch version framerd-3.9.1-29-ge3a33fa

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 10:58:54 -0400

framerd (3.9.1-11) unstable; urgency=low

  * More adjustments to libmysqlclient dependencies
  * git patch version framerd-3.9.1-25-g880479a

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 10:12:25 +0000

framerd (3.9.1-10) unstable; urgency=low

  * Fix libmysqlclient dependency
  * Save alternate booksystem files in config directory
  * git patch version framerd-3.9.1-23-g4def5c8

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 09:17:44 +0000

framerd (3.9.1-9) unstable; urgency=low

  * Add libzip2 dependency to packaging
  * Fix FramerD systemd configuration
  * git patch version framerd-3.9.1-18

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 14:47:16 +0000

framerd (3.9.1-8) unstable; urgency=low

  * Add optional arg to empty-string? to treat non-breaking spaces as content
  * git patch version framerd-3.9.1-15-gff06aeb

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 03:51:54 +0000

framerd (3.9.1-7) unstable; urgency=low

  * Fix bug in optimize code
  * git patch version framerd-3.9.1-12-g99c9b27

 -- Ken Haase <kh@beingmeta.com>  Sat, 30 Apr 2016 22:51:01 +0000

framerd (3.9.1-6) unstable; urgency=medium

  * Repackage for alternate libmagickwand under xenial
  * git patch version framerd-3.9.1-9-gad2c5fc

 -- Ken Haase <kh@beingmeta.com>  Fri, 29 Apr 2016 13:16:58 -0400

framerd (3.9.1-5) unstable; urgency=medium

  * Improve error capture and handling including interrupts
  * Add fdstatus and fdlog commands
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Thu, 31 Mar 2016 07:01:55 -0400

framerd (3.9.0-4) unstable; urgency=medium

  * Fix build flags to remove needless dependencies
  * git patch version framerd-3.9-15-ge4f2745

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 17:35:42 -0400

framerd (3.9.0-3) unstable; urgency=low

  * Fix opcode handling bugs
  * git patch version framerd-3.9-6-gc9661e4

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 11:14:47 -0400

framerd (3.9.0-2) unstable; urgency=low

  * Fix git tag and executable heralds
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 21:13:27 -0400

framerd (3.9.0-1) unstable; urgency=low

  * Fix systemd scripts
  * Added fdstatus script
  * Fix pool-load to set error context
  * Fixes to opcode bugs
  * Bumped minor release number

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 16:55:24 -0400

framerd (3.8.4-63) unstable; urgency=low

  * Fix libbson/libmongoc dependencies
  * git patch verison framerd-3.8.4-177-gdb5c379

 -- Ken Haase <kh@beingmeta.com>  Sun, 27 Mar 2016 00:31:54 +0000

framerd (3.8.4-62) unstable; urgency=low

  * Fix tidy5 dependencies
  * git patch version framerd-3.8.4-176-g33eb981

 -- Ken Haase <kh@beingmeta.com>  Sat, 26 Mar 2016 14:25:11 +0000

framerd (3.8.4-61) unstable; urgency=low

  * Add smarter immediate type checking
  * Fix some opcode/optimize bugs
  * git patch version framerd-3.8.4-173-gdc93e81

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 21:33:36 -0400

framerd (3.8.4-60) unstable; urgency=low

  * Fixes to xmlparsing
  * Fix some applytest error cases (whoops)
  * Check irritant pointer when calling fd_err
  * Improve backtrace CSS
  * git patch version framerd-3.8.4-168-g8a0394e

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 18:21:37 -0400

framerd (3.8.4-59) unstable; urgency=low

  * Add handling of extpool lockfn
  * make DOM cleanup take more initiative
  * git patch version framerd-3.8.4-156-g43cb8ea

 -- Ken Haase <kh@beingmeta.com>  Sat, 19 Mar 2016 19:51:14 +0000

framerd (3.8.4-58) unstable; urgency=low

  * Add set-procedure-args!
  * Add functions for augmenting errors in ONERROR handlers
  * Fixes to optimize module
  * Wrap primitives execution in call contours to convert setmped errors
    to return errors
  * git patch version framerd-3.8.4-153-gcd07c91

 -- Ken Haase <kh@beingmeta.com>  Mon, 14 Mar 2016 18:26:35 -0400

framerd (3.8.4-57) unstable; urgency=low

  * Pass priority on to logger from logif+
  * git patch version framerd-3.8.4-147-gfb67432

 -- Ken Haase <kh@beingmeta.com>  Sat, 05 Mar 2016 22:12:07 -0500

framerd (3.8.4-56) unstable; urgency=low

  * Fix bug in JWT changes
  * git patch version framerd-3.8.4-145-gebeeed4

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 10:43:32 -0500

framerd (3.8.4-55) unstable; urgency=low

  * Add RSA algorithm to JWT
  * Improve and fix parts of JWT module
  * Add third arg (urish) to ->base64 functions to use - and _
    for + and /
  * git patch version framerd-3.8.4-143-gdb7869b

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 08:57:17 -0500

framerd (3.8.4-54) unstable; urgency=low

  * Record boot handler and status in FramerD run dir
  * git patch version framerd-3.8.4-132-g22819cc (pre commit)

 -- Ken Haase <kh@beingmeta.com>  Mon, 29 Feb 2016 23:08:03 +0000

framerd (3.8.4-53) unstable; urgency=low

  * Fixes to daemon/servlet management scripts
  * git patch version framerd-3.8.4-128

 -- Ken Haase <kh@beingmeta.com>  Sun, 28 Feb 2016 20:09:55 +0000

framerd (3.8.4-52) unstable; urgency=low

  * Fix set-cookie to not automatically add dot to cookie domains
  * Fixes and refinements to jwt and jwt/auth
  * Make parsetime handle non-string arguments
  * git patch version framerd-3.8.4-126

 -- Ken Haase <kh@beingmeta.com>  Fri, 26 Feb 2016 22:58:38 +0000

framerd (3.8.4-51) unstable; urgency=low

  * Remove uridecode handling of '+' escapes
  * git patch version framer-3.8.4-121

 -- Ken Haase <kh@beingmeta.com>  Tue, 23 Feb 2016 18:51:41 +0000


etc/fileinfo.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/stat.h>

#define BLOCK_SIZE 8192

/*
 * sha1.c
 *
 * Originally witten by Steve Reid <steve@edmweb.com>
 * 
 * Modified by Aaron D. Gifford <agifford@infowest.com>
 *
 * NO COPYRIGHT - THIS IS 100% IN THE PUBLIC DOMAIN
 *
 * The original unmodified version is available at:
 *    ftp://ftp.funet.fi/pub/crypt/hash/sha/sha1.c
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * sha.h
 *
 * Originally taken from the public domain SHA1 implementation
 * written by by Steve Reid <steve@edmweb.com>
 * 
 * Modified by Aaron D. Gifford <agifford@infowest.com>
 *
 * NO COPYRIGHT - THIS IS 100% IN THE PUBLIC DOMAIN
 *
 * The original unmodified version is available at:
 *    ftp://ftp.funet.fi/pub/crypt/hash/sha/sha1.c
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/* Define this if your machine is LITTLE_ENDIAN, otherwise #undef it: */
#ifndef LITTLE_ENDIAN
#define LITTLE_ENDIAN 1
#endif

/* Make sure you define these types for your architecture: */
typedef unsigned int sha1_quadbyte;	/* 4 byte type */
typedef unsigned char sha1_byte;	/* single byte type */

/*
 * Be sure to get the above definitions right.  For instance, on my
 * x86 based FreeBSD box, I define LITTLE_ENDIAN and use the type
 * "unsigned long" for the quadbyte.  On FreeBSD on the Alpha, however,
 * while I still use LITTLE_ENDIAN, I must define the quadbyte type
 * as "unsigned int" instead.
 */

#define SHA1_BLOCK_LENGTH	64
#define SHA1_DIGEST_LENGTH	20

/* The SHA1 structure: */
typedef struct _SHA_CTX {
	sha1_quadbyte	state[5];
	sha1_quadbyte	count[2];
	sha1_byte	buffer[SHA1_BLOCK_LENGTH];
} SHA_CTX;

#ifndef NOPROTO
void SHA1_Init(SHA_CTX *context);
void SHA1_Update(SHA_CTX *context, sha1_byte *data, unsigned int len);
void SHA1_Final(sha1_byte digest[SHA1_DIGEST_LENGTH], SHA_CTX* context);
#else
void SHA1_Init();
void SHA1_Update();
void SHA1_Final();
#endif

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */

#ifdef LITTLE_ENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&(sha1_quadbyte)0xFF00FF00) \
	|(rol(block->l[i],8)&(sha1_quadbyte)0x00FF00FF))
#else
#define blk0(i) block->l[i]
#endif

#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
	^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);

typedef union _BYTE64QUAD16 {
	sha1_byte c[64];
	sha1_quadbyte l[16];
} BYTE64QUAD16;

/* Hash a single 512-bit block. This is the core of the algorithm. */
void SHA1_Transform(sha1_quadbyte state[5], sha1_byte buffer[64]) {
	sha1_quadbyte	a, b, c, d, e;
	BYTE64QUAD16	*block;

	block = (BYTE64QUAD16*)buffer;
	/* Copy context->state[] to working vars */
	a = state[0];
	b = state[1];
	c = state[2];
	d = state[3];
	e = state[4];
	/* 4 rounds of 20 operations each. Loop unrolled. */
	R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
	R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
	R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
	R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
	R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
	R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
	R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
	R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
	R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
	R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
	R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
	R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
	R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
	R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
	R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
	R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
	R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
	R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
	R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
	R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);
	/* Add the working vars back into context.state[] */
	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;
	state[4] += e;
	/* Wipe variables */
	a = b = c = d = e = 0;
}


/* SHA1_Init - Initialize new context */
void SHA1_Init(SHA_CTX* context) {
	/* SHA1 initialization constants */
	context->state[0] = 0x67452301;
	context->state[1] = 0xEFCDAB89;
	context->state[2] = 0x98BADCFE;
	context->state[3] = 0x10325476;
	context->state[4] = 0xC3D2E1F0;
	context->count[0] = context->count[1] = 0;
}

/* Run your data through this. */
void SHA1_Update(SHA_CTX *context, sha1_byte *data, unsigned int len) {
	unsigned int	i, j;

	j = (context->count[0] >> 3) & 63;
	if ((context->count[0] += len << 3) < (len << 3)) context->count[1]++;
	context->count[1] += (len >> 29);
	if ((j + len) > 63) {
	    memcpy(&context->buffer[j], data, (i = 64-j));
	    SHA1_Transform(context->state, context->buffer);
	    for ( ; i + 63 < len; i += 64) {
	        SHA1_Transform(context->state, &data[i]);
	    }
	    j = 0;
	}
	else i = 0;
	memcpy(&context->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */
void SHA1_Final(sha1_byte digest[SHA1_DIGEST_LENGTH], SHA_CTX *context) {
	sha1_quadbyte	i, j;
	sha1_byte	finalcount[8];

	for (i = 0; i < 8; i++) {
	    finalcount[i] = (sha1_byte)((context->count[(i >= 4 ? 0 : 1)]
	     >> ((3-(i & 3)) * 8) ) & 255);  /* Endian independent */
	}
	SHA1_Update(context, (sha1_byte *)"\200", 1);
	while ((context->count[0] & 504) != 448) {
	    SHA1_Update(context, (sha1_byte *)"\0", 1);
	}
	/* Should cause a SHA1_Transform() */
	SHA1_Update(context, finalcount, 8);
	for (i = 0; i < SHA1_DIGEST_LENGTH; i++) {
	    digest[i] = (sha1_byte)
	     ((context->state[i>>2] >> ((3-(i & 3)) * 8) ) & 255);
	}
	/* Wipe variables */
	i = j = 0;
	memset(context->buffer, 0, SHA1_BLOCK_LENGTH);
	memset(context->state, 0, SHA1_DIGEST_LENGTH);
	memset(context->count, 0, 8);
	memset(&finalcount, 0, 8);
}

static char hexdigits[16]="0123456789ABCDEF";

int main(int argc,char *argv[])
{
  if (argc<2) {
    fprintf(stderr,"Usage: fileinfo 'filename' [root]\n");
    exit(1);}
  else {
    char *filename=argv[1], *root=((argc<3)?(NULL):(argv[2]));
    char *abspath=NULL, *relpath=NULL;
    struct stat info; time_t mtime; struct tm *modtime, modtime_buf;
    unsigned char *content, *ptr, *original;
    int retval=stat(filename,&info);
    int i=0, file_len, bytes_read=0, abs_len;
    SHA_CTX context;
    sha1_byte digest[SHA1_DIGEST_LENGTH];
    unsigned char digest_string[41];
    FILE *f;
    if (retval<0) {
      fprintf(stderr,"STAT error %s\n",strerror(errno));
      exit(1);}
    else if (!(f=fopen(filename,"r"))) {
      fprintf(stderr,"FOPEN error %s\n",strerror(errno));
      exit(1);}
    mtime=info.st_mtime;
    memset(&modtime_buf,0,sizeof(struct tm));
    memset(digest,0,sizeof(digest));
    modtime=gmtime_r(&mtime,&modtime_buf);
    if ((*filename=='.')&&(filename[1]=='/')) filename=filename+2;
    if (*filename=='/') {
      int name_len=strlen(filename);
      abspath=malloc(name_len+1);
      strncpy(abspath,filename,name_len);
      abspath[name_len]='\0';
      abs_len=name_len;}
    else if ((root)&&(strcmp(root,".")==0)) {
      int name_len=strlen(filename);
      abspath=malloc(name_len+1);
      strncpy(abspath,filename,name_len);
      abspath[name_len]='\0';
      abs_len=name_len;}
    else {
      char *cwd=getcwd(NULL,1024);
      int cwd_len=strlen(cwd), name_len=strlen(filename);
      if (!(cwd)) {
	fprintf(stderr,"GETCWD error %s\n",strerror(errno));
	exit(1);}
      abspath=malloc(cwd_len+name_len+2);
      strncpy(abspath,cwd,cwd_len);
      if (cwd[cwd_len-1]!='/') {
	abspath[cwd_len]='/';
	strncpy(abspath+cwd_len+1,filename,name_len);
	abs_len=cwd_len+1+name_len;
      	abspath[abs_len]='\0';}
      else {
	strncpy(abspath+cwd_len,filename,name_len);
	abs_len=cwd_len+name_len;	
	abspath[abs_len]='\0';}
      if (cwd) free(cwd);}
    relpath=abspath;
    if (root) {
      char *scan=root;
      while ((*scan)&&((scan-root)<abs_len)) {
	if (*scan==*relpath) {scan++; relpath++;}
	else break;}
      if (*relpath=='\0') relpath=abspath;}
    file_len=info.st_size;
    ptr=content=original=(unsigned char *)malloc(file_len+1);
    while (bytes_read<file_len) {
      int delta=fread(ptr,sizeof(unsigned char),BLOCK_SIZE,f);
      if (delta<0) {
	fprintf(stderr,"FREAD error %s\n",strerror(errno));
	free(content); fclose(f);
	exit(1);}
      ptr=ptr+delta; bytes_read=bytes_read+delta;}
    SHA1_Init(&context);
    SHA1_Update(&context,content,file_len);
    SHA1_Final(digest,&context);
    i=0; while (i<20) {
      int byte=digest[i];
      digest_string[2*i]=hexdigits[(byte>>4)&0xF];
      digest_string[(2*i)+1]=hexdigits[byte&0xF];
      i++;}
    digest_string[40]='\0';
    fprintf(stdout,"%s %s %d-%02d-%02dT%02d:%02d:%02d\n",
	    relpath,digest_string,
	    1900+modtime->tm_year,modtime->tm_mon+1,
	    modtime->tm_mday,
	    modtime->tm_hour,modtime->tm_min,
	    modtime->tm_sec);
    fclose(f);
    free(original);
    return 0;}
}
